---
title: "Using schex with Seurat"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Seurat_schex}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  
---

```{r, include = FALSE}
library(here)
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = here::here("misc/images_Seurat/")
)
library(ggplot2)
theme_set(theme_classic())
```

Reduced dimension plotting is one of the essential tools for the analysis of 
single cell data. However, as the number of cells/nuclei in these these plots 
increases, the usefulness of these plots decreases. Many cells are plotted
on top of each other obscuring information, even when taking advantage of
transparency settings. This package provides binning strategies of cells/nuclei
into hexagon cells. Plotting summarized information of all cells/nuclei in their 
respective hexagon cells presents information without obstructions. The
package seemlessly works with the two most common object classes for the storage
of single cell data; `SingleCellExperiment` from the 
[SingleCellExperiment](https://bioconductor.org/packages/3.9/bioc/html/SingleCellExperiment.html) 
package and `Seurat` from the [Seurat](https://satijalab.org/seurat/) package. In
this vignette I will be presenting the use of `schex` for `Seurat` objects.

## Load libraries

```{r load-libraries, message=FALSE, warning=FALSE}
library(schex)
library(dplyr)
library(Seurat)
```

## Setup single cell data

In order to demonstrate the capabilities of the schex package, I will use the
a dataset of Peripheral Blood Mononuclear Cells (PBMC) freely available from 
10x Genomics. There are 2,700 single cells that were sequenced on the 
Illumina NextSeq 500. You can download the data from the 
[Seurat website](https://s3-us-west-2.amazonaws.com/10x.files/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz). 

```{r load, eval=FALSE}
pbmc.data <- Read10X(data.dir = here::here("misc/data/pbmc3k/filtered_gene_bc_matrices/hg19/"))
# Initialize the Seurat object with the raw (non-normalized data).
pbmc <- CreateSeuratObject(counts = pbmc.data, project = "pbmc3k", min.cells = 3, min.features = 200)
```

In the next few sections, I will perform some simple quality control steps
outlined in the [Seurat vignette](https://satijalab.org/seurat/v3.0/pbmc3k_tutorial.html). I will then calculate various dimension 
reductions and cluster the data also outlined in the vignette.

## Standard pre-processing workflow

### Filtering

Cells with high mitochondrial content as well as cells with too low
or too high feature count are filtered.

```{r filter, eval=FALSE}
pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")
pbmc <- subset(pbmc, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)
```

### Normalization

Next a global-scaling normalization method is employed to normalizes the 
feature expression measurements for each cell.

```{r norm, message=FALSE, warning=FALSE, eval=FALSE}
pbmc <- NormalizeData(pbmc, normalization.method = "LogNormalize", 
                      scale.factor = 10000, verbose=FALSE)
```

### Identification of highly variable genes

Many of the downstream methods are based on only the highly variable genes, hence
we require their identification.

```{r highly-variable, message=FALSE, warning=FALSE, eval=FALSE}
pbmc <- FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 2000,
                             verbose = FALSE)
```

### Scaling

Prior to dimension reduction the data is scaled.

```{r scaling, message=FALSE, warning=FALSE, eval=FALSE}
all.genes <- rownames(pbmc)
pbmc <- ScaleData(pbmc, features = all.genes, verbose = FALSE)
```

### Perform dimensionality reductions

First a PCA is applied to the data. Using the PCA you will have to decide on the
dimensionality of the data. Here the dimensionality was decided to be 10. 
Please refer to the original Seurat vignette for methods on how this is assessed.

```{r pca, message=FALSE, warning=FALSE, eval=FALSE}
pbmc <- RunPCA(pbmc, features = VariableFeatures(object = pbmc), 
               verbose = FALSE)
```

Next a UMAP dimensionality reduction is also run.

```{r umap, message=FALSE, warning=FALSE, eval=FALSE}
pbmc <- RunUMAP(pbmc, dims = 1:10, verbose=FALSE)
```

### Clustering

Finally the data is clustered.

```{r message=FALSE, warning=FALSE, eval=FALSE}
pbmc <- FindNeighbors(pbmc, dims = 1:10, verbose = FALSE)
pbmc <- FindClusters(pbmc, resolution = 0.5, verbose = FALSE)
```

## Plotting single cell data

At this stage in the workflow we usually would like to plot aspects of our data
in one of the reduced dimension representations. Instead of plotting this in an
ordinary fashion, I will demonstrate how schex can provide a better way of 
plotting this.

#### Calculate hexagon cell representation

First, I will calculate the hexagon cell representation for each cell for
a specified dimension reduction representation. I decide to use `nbins=35` which
specifies that I divide my x range into 40 bins. Note that this might be a 
parameter that you want to play around with depending on the number of cells/
nuclei in your dataset. Generally, for more cells/nuclei, `nbins` should be 
increased.

```{r calc-hexbin, eval=FALSE}
pbmc <- make_hexbin(pbmc, nbins = 35, 
                           dimension_reduction = "UMAP")
```

#### Plot number of cells/nuclei in each hexagon cell

First I plot how many cells are in each hexagon cell. This should be
relatively even, otherwise change the `nbins` parameter in the previous 
calculation.

```{r plot-density, fig.height=7, fig.width=7, eval=FALSE}
plot_hexbin_density(pbmc)
```

![](../misc/images_Seurat/plot-density-1.png)

#### Plot meta data in hexagon cell representation

Next I colour the hexagon cells by some meta information, such as the median 
total count or cluster membership in each hexagon cell.

```{r plot-meta-1, fig.height=7, fig.width=7, eval=FALSE}
plot_hexbin_meta(pbmc, col="nCount_RNA", action="median")
```

![](../misc/images_Seurat/plot-meta-1.png)
```{r plot-meta-2, fig.height=7, fig.width=7, eval=FALSE}
plot_hexbin_meta(pbmc, col="RNA_snn_res.0.5", action="majority")
```

![](../misc/images_Seurat/plot-meta-2.png)

For convenience there is also a function that allows the calculation of label
positions for factor variables. These can be overlayed with the package 
`ggrepel`.

```{r plot-meta-label, message=FALSE, fig.height=7, fig.width=7, eval=FALSE}
label_df <- make_hexbin_label(pbmc, col="RNA_snn_res.0.5")
pp <- plot_hexbin_meta(pbmc, col="RNA_snn_res.0.5", action="majority") 
pp + ggrepel::geom_label_repel(data = label_df, aes(x=x, y=y, label = label), 
                            colour="black",  label.size = NA, fill = NA)  
```

![](../misc/images_Seurat/plot-meta-label-1.png)

#### Plot gene expression in hexagon cell representation

Finally, I will visualize the gene expression of the CD19 gene in the 
hexagon cell representation.

```{r plot-gene, fig.height=7, fig.width=7, eval=FALSE}
gene_id <-"CD19"
plot_hexbin_gene(pbmc, type="logcounts", gene=gene_id, 
                 action="mean", xlab="UMAP1", ylab="UMAP2", 
                 title=paste0("Mean of ", gene_id))
```

![](../misc/images_Seurat/plot-gene-1.png)


